proposals/0293-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -175,161 +201,188 @@ struct MyView: View {
+The compiler will synthesize computed `text` and `$text` variables in the body of `insert`:
 
-* If the next `wrappedValue` accessor is `nonmutating`, then the mutability of the composed accessor is the same as the previous composed getter. If the wrapper is a reference type, the accessor is considered `nonmutating`.
-* If the next `wrappedValue` accessor is `mutating`, then the composed accessor is `mutating` if the previous composed getter _or_ setter is `mutating`, since both are needed to perform a writeback cycle.
+```swift
+func insert(text _text: Validated<String?>) { 
+  var text: String? {
+    get { _text.wrappedValue }
+  }
 
-If any of the property wrappers do not define a `wrappedValue` setter, then the wrapped property/parameter does not have a setter.
+  var $text: Result<String, ValidationResult> {
+    get { _text.projectedValue }
+  }
 
-### Call-site transformation
+  ...
+}
+```
 
-When passing an argument to a function with a property-wrapper parameter using the original argument label with no prefix, the compiler will wrap the argument in a call to `init(wrappedValue:)`. This transformation does _not_ apply to closures, because closures are not called with argument labels.
+### Call-site semantics
 
-#### Overload resolution of `init(wrappedValue)`
+When passing an argument to a function with a property-wrapped parameter, the compiler will wrap the argument in a call to the appropriate initializer depending on the argument label. When using the original argument label (or no argument label), the compiler will wrap the argument in a call to `init(wrappedValue:)`. When using the argument label prefixed with `$` (or `$_` in the case of no argument label), the compiler will wrap the argument in a call to `init(projectedValue:)`. 
 
-Since the property wrapper is initialized at the call-site, this means that the argument type can impact overload resolution of `init(wrappedValue:)`. For example, if a property wrapper defines overloads of `init(wrappedValue:)` with different generic constraints and that wrapper is used on a function parameter, e.g.:
+Consider the following property wrapper that implements both `init(wrappedValue:)` and `init(projectedValue:)`:
 
 ```swift
+struct Projection<Value> { ... }
+
 @propertyWrapper
-struct Wrapper<Value> {
+public struct Wrapper<Value> {
 
-  init(wrappedValue: Value) { ... }
+  public init(wrappedValue: Value) { ... }
 
-  init(wrappedValue: Value) where Value : Collection { ... }
-  
-}
+  public init(projectedValue: Projection<Value>) { ... }
 
+  public var wrappedValue: Value
 
-func generic<T>(@Wrapper arg: T) { ... }
+  public var projectedValue: Projection<Value> { ... }
+
+}
 ```
 
-Then, overload resolution will choose which `init(wrappedValue:)` to call based on the static type of the argument at the call-site:
+A function with a `@Wrapped` parameter can be called with either a wrapped value or a projected value:
```

Comment:
There is a typo here, the property wrapper is called `struct Wrapper` but the line talks about `@Wrapped`, fix:
```suggestion
A function with a `@Wrapper` parameter can be called with either a wrapped value or a projected value:
```

Reply:
Thanks for pointing this out! This example is mostly a placeholder, and I'll replace it when @filip-sakel completes the Motivation section with new examples for this design.